validator_client/src/initialized_validators.rs

```diff
@@ -611,4 +657,81 @@ impl InitializedValidators {
         }
         Ok(())
     }
+
+    /// Returns the epoch a validator should perform doppelganger detection until, starting
+    /// from when this method is called.
+    ///
+    /// This will return `None` if the current epoch is the genesis epoch or if we are prior to genesis.
+    /// This is because we cannot perform doppelganger detection in the first epoch after genesis
+    /// or we risk having no validators to propose blocks.
+    fn determine_doppelganger_detection_epoch(
+        &self,
+        current_epoch: Epoch,
+        genesis_epoch: Epoch,
+    ) -> Option<Epoch> {
+        (current_epoch > genesis_epoch && !self.disable_doppelganger_detection)
+            .then(|| current_epoch + DOPPELGANGER_DETECTION_EPOCHS)
+    }
+
+    /// Updates the doppelganger detection epoch of all initialized validators based on the provided
+    /// epoch.
+    pub fn update_all_doppelganger_detection_epochs(
+        &mut self,
+        current_epoch: Epoch,
+        genesis_epoch: Epoch,
+    ) {
+        let detection_epoch =
+            self.determine_doppelganger_detection_epoch(current_epoch, genesis_epoch);
+        for (_, val) in self.validators.iter_mut() {
+            val.doppelganger_detection_epoch = detection_epoch;
+        }
+    }
+
+    /// Gets the public key of all validators who are not in the doppelganger detection period.
+    pub fn iter_signing_pubkeys(
+        &self,
+        current_epoch: Epoch,
+    ) -> impl Iterator<Item = &PublicKeyBytes> {
+        self.validators.iter().filter_map(move |(pubkey, val)| {
+            val.doppelganger_detection_epoch
+                .map_or(true, |doppelganger_epoch| {
+                    doppelganger_epoch < current_epoch
+                })
+                .then(|| pubkey)
+        })
+    }
+
+    /// Gets all validators that are in the doppelganger detection period in the given epoch.
+    pub fn get_doppelganger_detecting_validators(&self, epoch: Epoch) -> Vec<ValidatorId> {
+        self.validators
+            .iter()
+            .filter_map(|(_, val)| {
+                // make sure we've determined this validator exists in the beacon chain
+                match val.index {
+                    Some(index) => {
+                        val.doppelganger_detection_epoch
+                            .map(|doppelganger_epoch| {
+                                // We want to avoid checking the epoch in which doppelganger detection was started
+                                // so we don't pick up attestations from our own validator on restart.
+                                (doppelganger_epoch >= epoch
+                                    && epoch != doppelganger_epoch - DOPPELGANGER_DETECTION_EPOCHS)
```

Comment:
Nitpick: I'm starting to go cold on relying on the saturating methods of slot/epoch since a seemingly innocent change of type to `u64` can cause unexpected consequences.

```suggestion
                                    && epoch != doppelganger_epoch.saturating_sub(DOPPELGANGER_DETECTION_EPOCHS))
```